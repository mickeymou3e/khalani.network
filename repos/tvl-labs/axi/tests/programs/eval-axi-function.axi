function factorial(n: Nat) -> Nat {
    match n {
        Zero => 1,
        Succ(m) => n mul factorial(m),
    }
}

!prove_by_eval_term(factorial(0));
!prove_by_eval_term(factorial(1));
!prove_by_eval_term(factorial(5));

function fibonacci(n: Nat) -> Nat {
    match n {
        Succ(Zero) => 1,
        Succ(Succ(m)) => fibonacci(m) add fibonacci(m add 1),
        Zero => 0,
    }
}

!prove_by_eval(fibonacci(0) = 0);
!prove_by_eval(fibonacci(1) = 1);
!prove_by_eval(fibonacci(8) = 21);

inductive ListNat {
    ListNatNil,
    ListNatCons(Nat, ListNat),
}

function sum_list_nat(xs: ListNat) -> Nat {
    let ys := xs;
    let zero := Zero;

    match ys {
        ListNatNil => zero,
        ListNatCons(x, xs) => x add sum_list_nat(xs),
    }
}

!prove_by_eval(sum_list_nat(
    ListNatCons(
        1,
        ListNatCons(
            2,
            ListNatCons(
                3,
                ListNatNil,
            ),
        ),
    )
) = 6);

function my_predicate(x: Nat) -> Boolean {
    x lt 3 | x = 3
}

!prove_by_eval(my_predicate(1));
!prove_by_eval(my_predicate(3));
!prove_by_eval(my_predicate(Succ(Succ(Succ(Zero)))));

function my_boolean_function() -> Boolean {
    my_predicate(1) | false
}

!prove_by_eval(my_boolean_function());

# Should fail.
!prove_by_eval(my_predicate(4));
