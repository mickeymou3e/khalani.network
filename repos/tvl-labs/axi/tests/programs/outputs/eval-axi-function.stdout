Result: Nothing
Result: Proved(=[factorial[0]):Nat, 1])
Result: Proved(=[factorial[1]):Nat, 1])
Result: Proved(=[factorial[5]):Nat, 120])
Result: Nothing
Result: Proved(=[fibonacci[0]):Nat, 0])
Result: Proved(=[fibonacci[1]):Nat, 1])
Result: Proved(=[fibonacci[8]):Nat, 21])
Result: Nothing
Result: Nothing
Result: Proved(=[sum_list_nat[ListNatCons[1, ListNatCons[2, ListNatCons[3, ListNatNil:ListNat]):ListNat]):ListNat]):ListNat]):Nat, 6])
Result: Nothing
Result: Proved(my_predicate[1])
Result: Proved(my_predicate[3])
Result: Proved(my_predicate[Succ[Succ[Succ[Zero:Nat]):Nat]):Nat]):Nat])
Result: Nothing
Result: Proved(my_boolean_function[])
